src/envs/coffee.py

```diff
@@ -0,0 +1,756 @@
+        robot, jug = objects
+        return self._Holding_holds(state, [robot, jug])
+
+    def _PlaceJugInMachine_policy(self, state: State, memory: Dict,
+                                  objects: Sequence[Object],
+                                  params: Array) -> Action:
+        # This policy picks the jug up slightly above the table to avoid
+        # worrying about friction, then moves directly to the place position,
+        # then places the jug.
+        del memory, params  # unused
+        robot, jug, _ = objects
+        # Use the jug position as the origin.
+        x = state.get(jug, "x")
+        y = state.get(jug, "y")
+        z = state.get(robot, "z") - self.jug_handle_height
+        jug_pos = (x, y, z)
+        place_pos = (self.dispense_area_x, self.dispense_area_y, self.z_lb)
+        # If close enough, place.
+        sq_dist_to_place = np.sum(np.subtract(jug_pos, place_pos)**2)
+        if sq_dist_to_place < self.dispense_tol:
+            return Action(np.array([0.0, 0.0, 0.0, 0.0, 1.0],
+                                   dtype=np.float32))
+        # If already above the table, move directly toward the place pos.
+        if z > self.z_lb:
+            return self._get_move_action(place_pos, jug_pos)
+        # Move up.
+        return self._get_move_action((x, y, z + self.max_position_vel),
+                                     jug_pos)
+
+    def _PlaceJugInMachine_terminal(self, state: State, memory: Dict,
+                                    objects: Sequence[Object],
+                                    params: Array) -> bool:
+        del memory, params  # unused
+        robot, jug, machine = objects
+        return not self._Holding_holds(state, [robot, jug]) and \
+            self._InMachine_holds(state, [jug, machine])
+
+    def _TurnMachineOn_policy(self, state: State, memory: Dict,
+                              objects: Sequence[Object],
+                              params: Array) -> Action:
+        # This policy moves the robot up to be level with the button in the
+        # z direction and then moves forward in the y direction to press it.
+        del memory, params  # unused
+        robot, _ = objects
+        x = state.get(robot, "x")
+        y = state.get(robot, "y")
+        z = state.get(robot, "z")
+        robot_pos = (x, y, z)
+        button_pos = (self.button_x, self.button_y, self.button_z)
+        if (self.button_z - z)**2 < self.button_radius**2:
+            # Move directly toward the button.
+            return self._get_move_action(button_pos, robot_pos)
+        # Move only in the z direction.
+        return self._get_move_action((x, y, self.button_z), robot_pos)
+
+    def _TurnMachineOn_terminal(self, state: State, memory: Dict,
+                                objects: Sequence[Object],
+                                params: Array) -> bool:
+        del memory, params  # unused
+        _, machine = objects
+        return self._MachineOn_holds(state, [machine])
```

Comment:
why is the termination condition not PressingButton? actually in general i'm confused about the relationship between PressingButton and MachineOn

Reply:
PressingButton would work too. the button is pressed and the machine turns on at the same time. the difference is that MachineOn holds forever after, but PressingButton is deleted when the robot moves away.